src/lib/storage/reference_segment.cpp

```diff
@@ -0,0 +1,36 @@
+#include <unordered_set>
+
+#include "reference_segment.hpp"
+#include "type_cast.hpp"
+
+#include "utils/assert.hpp"
+
+namespace opossum {
+
+ReferenceSegment::ReferenceSegment(const std::shared_ptr<const Table>& referenced_table,
+                                   const ColumnID referenced_column_id, const std::shared_ptr<const PosList>& pos)
+    : _referenced_table(referenced_table), _referenced_column_id(referenced_column_id), _pos(pos) {}
+
+AllTypeVariant ReferenceSegment::operator[](const ChunkOffset chunk_offset) const {
+  const auto position = (*_pos)[chunk_offset];
+  const auto chunk_index = position.chunk_id;
```

Comment:
possibly rename to `chunk_id` for consistent naming

```suggestion
  const auto chunk_id = position.chunk_id;
```